pages/design_documentation/ocf_io_engines.md

```diff
@@ -0,0 +1,289 @@
+~~~~
+	[OCF_IO_WT_IF] = {
+		.read = ocf_read_generic,       
+                .write = ocf_write_wt,          
+                .name = "Write Through"
+	}
+~~~~
+
+### Write Through - write path
+A write request issued to the */dev/casX-X* file will be issued both to the core and
+cache devices simultaneously. 
+
+If the data in the cache is dirty or it is a cache miss, the respective metadata 
+bits are updated accordingly so the data in cache becomes valid and clean.
+
+In some less common circumstances, the data may be written using the 
+[**Pass Through**](/ocf_io_engines.html#pass-through-pt) engine, bypassing the cache
+device altogether. This may happen if it is a miss and logic in 
+*ocf_prepare_cachelines()* decides is not subject to insertion. 
+
+### Write Through - read path 
+A read request in this cache mode is handled by *ocf_read_generic* function.
+
+If the requested data exists in the cache - a cache hit situation - the data is 
+read from a cache drive.
+
+If the data is not completely clean (read partial dirty hit) and partially invalid,
+a [**cleaner**](/cleaning.html) is invoked to clean the data, i.e. flush the dirty
+data to the core. Afterwards, the request is handled again and the data is
+read from the core and returned.
+
+If the data does not exist in the cache (cache miss), the requested data is read 
+from the core device. Additionally, a **backfill** is performed in the background, 
+where the missing data will eventually be inserted into the cache to accelerate 
+subsequent reads.
+
+In case of a partial miss and if some target sectors are dirty (partial dirty miss),
+cleaning is performed first and then the data will be read from the core.
+
+
+## Write Back (WB)
+The write-back engine writes the data to a cache device only. The data will be 
+stored on a core device only after being cleaned - flushed to core device.
+This engine is used primarily in the Write Back cache mode to accelerate both 
+write and read operations at the cost of a higher risk of data loss if the cache 
+drive fails. 
+
+See [**write back cache mode configuration**](/cache_configuration.html#write-back)
+for details.
+	
+Write Back cache mode defines its read and write operations as follows:
+~~~~
+[OCF_IO_WB_IF] = {
+        .read = ocf_read_generic,
+        .write = ocf_write_wb,
+        .name = "Write Back"
+},      
+~~~~
+
+### Write Back - write path
+The requested data is written to the cache device only, producing dirty data.
+This data is then periodically cleaned by the [**cleaner**](/cleaning.html). 
+
+The request may be handled by [**Pass Through**](/ocf_io_engines.html#pass-through-pt) 
+engine if it cannot be handled by Write Back method. 
+See [**Write Through write path**](/ocf_io_engines.html#write-through---write-path).
+
+### Write Back - read path
+Read requests are handled by the generic read function, as in 
+[**Write Through read path**](/ocf_io_engines.html#write-through---read-path).
```

Comment:
Do we want to have description on WB cache mode as a subsection of wb engine documentation? It looks a bit upside down.

Reply:
again, apart maybe from wording (which btw. is not wrong - WB cache mode uses these engines, but that can be easily changed), the relation of these engines to the cache mode is clear and incidental with regard to this documentation paragraph, @arutk , what's your opinion?

Reply:
After discussion with @kmajzero we agreed that here we will describe engines in isolation, regardless of associated cache mode / io interface.